### PR TITLE
Revert two unapproved changes to rustc_typeck.

### PR DESCRIPTION
**Pull request by eddyb**
There was a breakdown in process (https://github.com/rust-lang/rust/pull/59004#issuecomment-477600735, https://github.com/rust-lang/rust/pull/58894#discussion_r272795560) and two changes were made to `rustc_typeck`'s "collect" queries, for rustdoc, that were neither needed *nor* correct.
I'm reverting them here, and will fix up rustdoc *somehow*, if necessary.

cc @rust-lang/compiler How do we ensure this doesn't happen again?

r? @nikomatsakis or @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
> cc @rust-lang/compiler How do we ensure this doesn't happen again?

We could setup bors to require a github approval from a relevant team member on certain source paths.

**Comment by petrochenkov** [comment]
>How do we ensure this doesn't happen again?

Sigh, resolve has a bunch of rustdoc-specific (and diagnostics-specific) hacks that don't work entirely correctly, but at least they don't affect the normal compiler run.
I ended up letting them live, not blocking PRs introducing them, and just hoping to remove or refactor them into something more correct and aesthetically pleasing some day.
If I had all the time in the world, I'd of course took every PR from GuillaumeGomez and rewrote it properly, but a man can dream...

**Comment by GuillaumeGomez** [comment]
> If I had all the time in the world, I'd of course took every PR from GuillaumeGomez and rewrote it properly, but a man can dream...

All of them have been approved except for the two last ones (and on the last one I did ping the compiler team!). If not happy with the PR, you can comment on it. By doing this, you'll reintroduce bugs in rustdoc. I'm against this changes. The process hasn't been followed, I'm sorry about that, but instead, wouldn't it be better to fix the librustc_typeck bad changes I made directly? As far as I can tell, it doesn't break anything in the rustc internals whereas reverting will break rustdoc again.

**Comment by petrochenkov** [comment]
@GuillaumeGomez 
(Sorry, I have my opinions accumulated over the years on which people tend to introduce hacks, but I shouldn't been publicly targeting you specifically.)

**Comment by oli-obk** [comment]
I opened https://github.com/rust-lang/rust/issues/59790 for further discussion. This PR should of course not break rustdoc, but address the original issues in a manner not affecting regular compilation.

**Comment by eddyb** [comment]
> If not happy with the PR, you can comment on it.

There should be an explicit review and approval, i.e. default to *not* merge non-rustc changes to rustc crates, instead of defaulting to merge in the absence of requested changes.

> As far as I can tell, it doesn't break anything in the rustc internals whereas reverting will break rustdoc again.

As I've explained in https://github.com/rust-lang/rust/issues/59790#issuecomment-480763327, this is *undocumented cruft*, and as such it will cause confusion or even stop some refactors, if nobody checks the change history and just assumes the code is needed.

And I *obviously* can't break `rustdoc`, since we have tests! So I'm forced to do a minimal fix, at least.

<hr/>

@petrochenkov FWIW, I think rustdoc as a whole needs some refactors, and @rust-lang/rustdoc has to make due with the codebase they've inherited. I've offered a few times but ended up with not enough spare time to do said refactors (although maybe I should try again, e.g. the comments I left on #58894).

Overall I'd rather stick the blame on the grandfathered codebase, and process breakdown, than individuals.

**Comment by QuietMisdreavus** [comment]
If https://github.com/rust-lang/rust/pull/59004 and https://github.com/rust-lang/rust/pull/58894 were the PRs that introduced the problematic changes, we can revert them wholesale. They were an enhancement and cosmetic fix, respectively, and reverting them should not introduce any show-stopping bugs to rustdoc.

It wouldn't hurt to get more eyes on rustdoc PRs more generally; i know i've been struggling to stay on top of my review queue, and having more hands would help keep everything going smoothly.

**Comment by eddyb** [comment]
@QuietMisdreavus no need, for #59004 I already have an equivalent replacement (since the data *is already* in `rustdoc`, no need to ask typeck to recompute it), while for #58894 I almost already have the proper solution implemented (exposing `explicit_predicates_of` cross-crate, which is mostly copy-paste).

**Comment by GuillaumeGomez** [comment]
@QuietMisdreavus Considering that #59004 is a big search improvement, it'd be quite the regression to revert. Luckily for me, it's not needed. 😌

**Comment by eddyb** [comment]
@rust-timer build b2e5868d670a1e0653d9708e49308b6629aea5f8

**Comment by eddyb** [comment]
cc @nikomatsakis @michaelwoerister @nnethercote How bad are the perf results?

I could try to supply `predicates_defined_on` (which is equivalent to `predicates_of` except for traits) from `rustc_metadata` by deserializing the `explicit_predicates_of` and then deserializing and appending the `inferred_outlives_of` *without* allocating either into an `Lrc`, but I'm not sure that's needed.

**Comment by nnethercote** [comment]
The worst-affected benchmarks are the short-running, less important ones. But it's still clearly a regression :(

**Comment by nnethercote** [comment]
Here is a Cachegrind diff for a "Check CleanIncr" build of `ripgrep`, which was one the biggest regressions in a larger benchmark.
```
--------------------------------------------------------------------------------
Ir                 
--------------------------------------------------------------------------------
9,063,571 (100.0%)  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir                   file:function
--------------------------------------------------------------------------------
-1,355,130 (-15.0%)  /home/njn/moz/rustN/src/libserialize/serialize.rs:serialize::serialize::Decoder::read_struct_field
   899,266 ( 9.92%)  /home/njn/moz/rustN/src/librustc_metadata/decoder.rs:rustc_metadata::decoder::DecodeContext::read_lazy_distance
   772,315 ( 8.52%)  /home/njn/moz/rustN/src/libserialize/serialize.rs:<rustc_metadata::schema::Entry as serialize::serialize::Decodable>::decode
   647,102 ( 7.14%)  /home/njn/moz/rustN/src/libstd/collections/hash/table.rs:rustc::dep_graph::prev::PreviousDepGraph::new
   575,554 ( 6.35%)  /home/njn/moz/rustN/src/libcore/ptr.rs:rustc::dep_graph::prev::PreviousDepGraph::new
   493,118 ( 5.44%)  /home/njn/moz/rustN/src/libstd/collections/hash/map.rs:rustc::dep_graph::prev::PreviousDepGraph::new
   431,526 ( 4.76%)  /home/njn/moz/rustN/src/librustc_metadata/decoder.rs:<rustc_metadata::schema::Entry as serialize::serialize::Decodable>::decode
  -408,336 (-4.51%)  /home/njn/moz/rustN/src/librustc_metadata/schema.rs:<rustc_metadata::schema::Entry as serialize::serialize::Decodable>::decode
   380,927 ( 4.20%)  /home/njn/moz/rustN/src/libcore/intrinsics.rs:rustc::dep_graph::prev::PreviousDepGraph::new
   379,621 ( 4.19%)  /home/njn/moz/rustN/src/libserialize/opaque.rs:<serialize::opaque::Decoder as serialize::serialize::Decoder>::read_usize
   338,776 ( 3.74%)  /home/njn/moz/rustN/src/libserialize/leb128.rs:rustc_incremental::persist::load::load_dep_graph::{{closure}}::{{closure}}
   301,719 ( 3.33%)  /home/njn/moz/rustN/src/librustc/dep_graph/graph.rs:rustc::dep_graph::graph::DepGraph::try_mark_previous_green
   293,040 ( 3.23%)  /home/njn/moz/rustN/src/libserialize/serialize.rs:<u32 as serialize::serialize::Encodable>::encode
   281,777 ( 3.11%)  /home/njn/moz/rustN/src/libserialize/leb128.rs:<u32 as serialize::serialize::Encodable>::encode
   276,688 ( 3.05%)  /home/njn/moz/rustN/src/libserialize/leb128.rs:<serialize::opaque::Decoder as serialize::serialize::Decoder>::read_usize
   245,283 ( 2.71%)  /home/njn/moz/rustN/src/libcore/num/mod.rs:rustc::dep_graph::prev::PreviousDepGraph::new
   217,059 ( 2.39%)  /home/njn/moz/rustN/src/libstd/collections/hash/map.rs:rustc::dep_graph::graph::CurrentDepGraph::intern_node
  -189,246 (-2.09%)  /home/njn/moz/rustN/src/librustc_metadata/decoder.rs:serialize::serialize::Decoder::read_struct_field
   183,414 ( 2.02%)  /home/njn/moz/rustN/src/librustc/dep_graph/graph.rs:rustc::dep_graph::graph::CurrentDepGraph::intern_node
   177,540 ( 1.96%)  /home/njn/.cargo/registry/src/github.com-1ecc6299db9ec823/smallvec-0.6.7/lib.rs:smallvec::SmallVec<A>::push
   169,388 ( 1.87%)  /home/njn/moz/rustN/src/liballoc/vec.rs:<u32 as serialize::serialize::Encodable>::encode
   151,618 ( 1.67%)  /home/njn/moz/rustN/src/libcore/slice/mod.rs:rustc::dep_graph::graph::DepGraph::try_mark_previous_green
   148,582 ( 1.64%)  /home/njn/moz/rustN/src/libcore/slice/mod.rs:rustc::dep_graph::graph::DepGraph::serialize
   129,970 ( 1.43%)  /home/njn/moz/rustN/src/libstd/collections/hash/table.rs:rustc_resolve::Resolver::set_binding_parent_module
   129,382 ( 1.43%)  /home/njn/moz/rustN/<::rustc::ty::codec::__impl_decoder_methods macros>:<rustc_metadata::schema::Entry as serialize::serialize::Decodable>::decode
   122,276 ( 1.35%)  /home/njn/moz/rustN/src/librustc_data_structures/fingerprint.rs:rustc_data_structures::fingerprint::Fingerprint::encode_opaque
   103,533 ( 1.14%)  /home/njn/moz/rustN/src/libcore/slice/mod.rs:<serialize::opaque::Decoder as serialize::serialize::Decoder>::read_usize
   100,119 ( 1.10%)  /home/njn/moz/rustN/src/librustc/ty/query/on_disk_cache.rs:rustc::ty::query::on_disk_cache::OnDiskCache::load_diagnostics
    99,423 ( 1.10%)  /home/njn/moz/rustN/src/libstd/collections/hash/table.rs:rustc::dep_graph::graph::CurrentDepGraph::intern_node
    93,258 ( 1.03%)  /home/njn/moz/rustN/src/libserialize/serialize.rs:rustc_incremental::persist::load::load_dep_graph::{{closure}}::{{closure}}
```

**Comment by eddyb** [comment]
@nnethercote Now the encoding/decoding should be the same as before in most cases.

**Comment by nnethercote** [comment]
The "Check CleanIncr" results for `ripgrep` were barely changed by the updated code.

**Comment by eddyb** [comment]
Ugh, I completely misunderstood the slowdown.
It's going into`rustc_metadata` *at all*. Every cross-crate query deserializes the `Entry` for a `DefId` over and over again, and that's not cheap...
I guess I'll look into ways of mitigating that, which should make this patch negligible.

**Comment by eddyb** [comment]
I've opened #59953 with the tentative performance fix, which should reduce the overhead of this PR.

**Comment by nikomatsakis** [comment]
@eddyb so is this PR blocked on #59953? what's the story here?

**Comment by eddyb** [comment]
@nikomatsakis Yeah, and that's waiting for a review from @michaelwoerister.

**Comment by eddyb** [comment]
@bors rollup=never (may cause perf regressions)

Also, this is now blocked on #65541.

**Comment by eddyb** [comment]
Note to self: once #65541 is merged, do one of these: https://github.com/rust-lang/rust/pull/59953#issuecomment-542521919.
(because of changes in metadata tables)

**Comment by GuillaumeGomez** [comment]
Is there anything I can do to help in order to keep https://github.com/rust-lang/rust/pull/59004 rustdoc changes? Like I said, it's a big rustdoc's search improvement so it'd be nice to keep it. :-/

**Comment by eddyb** [comment]
I think I left comments on the original PR on how to do it correctly without changing `rustc_typeck`, but I don't recall trying to implement it. I can take a stab if you want.

**Comment by eddyb** [comment]
@GuillaumeGomez I just checked [the first "Revert" commit of this PR](https://github.com/rust-lang/rust/pull/59789/commits/650f4bafbd5751908a1f3e69139cac57265f1f13) and it only reverts the changes to `rustc_typeck`, tweaking `rustdoc` so it remains working without needing said changes.

So there's nothing to worry about, I think?

**Comment by GuillaumeGomez** [comment]
If so then it's awesome! Sorry to bring it up then. :)

**Comment by eddyb** [comment]
| | Before (9285d401a6070094747465962bc49969b93e14c5) <br> [rust-std.tar.xz](https://rust-lang-ci2.s3-us-west-1.amazonaws.com/rustc-builds/9285d401a6070094747465962bc49969b93e14c5/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz) | After (0ba7f2605ec9d9c443d7d70bc109b69a79c0f79d) <br> [rust-std.tar.xz](https://rust-lang-ci2.s3-us-west-1.amazonaws.com/rustc-builds/0ba7f2605ec9d9c443d7d70bc109b69a79c0f79d/rust-std-nightly-x86_64-unknown-linux-gnu.tar.xz) | Increase |
| --- | --: | --: | --- |
| `libcore.rlib` | 26.54MB | 26.63MB | 0.3% |
| `libstd.rlib` | 10.37MB | 10.36MB | -0.1% |
| Total (`.tar.xz`) | 18.19MB | 18.67MB | 2.6% |

**Comment by JohnCSimon** [comment]
Ping from triage?
@eddyb Can you please address the merge conflict?
cc: @nikomatsakis 
Thank you!

**Review by nikomatsakis — APPROVED** [review]

**Comment by eddyb** [comment]
@bors retry
(spurious network failure)

**Pull request merged**
